<a href="https://colab.research.google.com/github/rutuser/03MAIR----Algoritmos-de-Optimizacion/blob/main/Actividad_Seminario_Algoritmos_Michal_Ruzicka_Ruzicka.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Algoritmos de optimización - Seminario<br>
Nombre y Apellidos: Michal Ruzicka Ruzicka  <br>
Url: https://github.com/.../03MAIR---Algoritmos-de-Optimizacion---2019/tree/master/SEMINARIO<br>
Problema:
> 1. Elección de grupos de población homogéneos <br>
>2. Organizar los horarios de partidos de La Liga<br>
>3. Combinar cifras y operaciones

Descripción del problema:(copiar enunciado)

### Problema 3. - Combinar cifras y operaciones <br>
• El problema consiste en analizar el siguiente problema y diseñar un algoritmo que lo resuelva. <br>
• Disponemos de las 9 cifras del 1 al 9 (excluimos el cero) y de los 4 signos básicos de las
operaciones fundamentales: suma(+), resta(-), multiplicación(*) y división(/) <br>
• Debemos combinarlos alternativamente sin repetir ninguno de ellos para obtener una
cantidad dada. Un ejemplo sería para obtener el 4:
`4+2-6/3x1`

Debe analizarse el problema para encontrar todos los valores enteros posibles planteando las
siguientes cuestiones: <br>
- ¿Qué valor máximo y mínimo se pueden obtener según las condiciones del problema? <br>
- ¿Es posible encontrar todos los valores enteros posibles entre dicho mínimo y máximo ? <br>
- Nota: Es posible usar la función de python `eval` para evaluar una expresión

(*) La respuesta es obligatoria





                                        

(*)¿Cuantas posibilidades hay sin tener en cuenta las restricciones?<br>



¿Cuantas posibilidades hay teniendo en cuenta todas las restricciones.




Respuesta

In [49]:
from itertools import permutations, product
import random

In [50]:
numbers = ['1', '2', '3', '4', '5', '6', '7', '8', '9']
operators = ['+', '-', '*', '/']

In [28]:
# Numero de combinaciones sin restricciones (con repeticiones de numeros y operaciones)

# Usando el producto cartesiano
combinaciones_numeros = list(product(numbers,repeat=len(operators)+1))

combinaciones_operadores = list(product(operators,repeat=len(operators)))

combinaciones_totales = len(combinaciones_numeros) * len(combinaciones_operadores)

combinaciones_totales

15116544

In [29]:
# Numero de combinaciones con restricciones (sin repeticiones de numeros y operaciones)

# Usando permutaciones
combinaciones_numeros = list(permutations(numbers,len(operators)+1))

combinaciones_operadores = list(permutations(operators,len(operators)))

combinaciones_totales = len(combinaciones_numeros) * len(combinaciones_operadores)

combinaciones_totales

362880

Modelo para el espacio de soluciones<br>
(*) ¿Cual es la estructura de datos que mejor se adapta al problema? Argumentalo.(Es posible que hayas elegido una al principio y veas la necesidad de cambiar, arguentalo)


Respuesta

#### La mejor estructura de datos que se adapta al problema es un `list()` o `set()`. Al principio pensé en usar un `set()`
#### para asegurarme de que las soluciones fuesen únicas, pero más tarde me di cuenta de que al usar permutaciones,
#### las soluciones ya serán únicas.

Según el modelo para el espacio de soluciones<br>
(*)¿Cual es la función objetivo?

(*)¿Es un problema de maximización o minimización?

Respuesta

#### Función objetivo: `n!, n = 9, 9! = 362880`
#### Problema de minimización

Diseña un algoritmo para resolver el problema por fuerza bruta

Respuesta

In [51]:
# Espacio de soluciones
espacio_soluciones = set()

for values in permutations(numbers,len(operators)+1):
    for opers in permutations(operators,len(operators)):
      formula = "".join(o+v for o,v in zip([""]+list(opers),values)) # Intercalar los numeros con los operadores
      evaluacion = eval(formula)

      if evaluacion - int(evaluacion) == 0: # Solo buscamos numeros enteros
        espacio_soluciones.add(eval(formula))

min_value = min(espacio_soluciones)
max_value = max(espacio_soluciones)

print('El espacio de soluciones se encuentra entre el ', min_value, 'y el', max_value)

El espacio de soluciones se encuentra entre el  -69.0 y el 77.0


In [52]:
def search_formula(num):
  resultado = list()
  counter = 0

  for values in permutations(numbers,len(operators)+1):
    for opers in permutations(operators,len(operators)):
      formula = "".join(o+v for o,v in zip([""]+list(opers),values)) # Intercalar los numeros con los operadores
      evaluacion = eval(formula)
      counter += 1

      if evaluacion == num:
        print(counter, 'iteraciones para el numero', str(num)+'.', 'Formula encontrada ->', formula)
        return [formula, counter]

In [55]:
# Elegimos un numero aleatorio de nuestro espacio de soluciones
# para tomar de referencia su resultado y comprobar si mejoramos el algoritmo mas tarde
rand_number = random.randint(min_value,max_value)

In [56]:
# Aplicamos nuestro algoritmo al numero elegido
resultado = search_formula(rand_number)

41397 iteraciones para el numero 52. Formula encontrada -> 2/1-4+6*9


Calcula la complejidad del algoritmo por fuerza bruta

Respuesta

#### La complejidad puede variar desde `O(1)` hasta `O(n*m)` dependiendo del número e iteracion en la que se encuentre a formula
#### Siendo n el número de permutaciones posibles de números, y m el número de permutaciones posibles de operadores
#### `n = permutations(numbers, len(operators)+1)`
#### `m = permutations(operators, len(operators))`

(*)Diseña un algoritmo que mejore la complejidad del algortimo por fuerza bruta. Argumenta porque crees que mejora el algoritmo por fuerza bruta

Respuesta

In [57]:
def search_formula_rand(num):
  resultado = list()
  counter = 0

  #Iterando de forma aleatoria.
  for values in sorted(permutations(numbers,len(operators)+1), key=lambda _: random.random()):
    for opers in sorted(permutations(operators,len(operators)), key=lambda _: random.random()):
      formula = "".join(o+v for o,v in zip([""]+list(opers),values)) # Intercalar los numeros con los operadores
      evaluacion = eval(formula)
      counter += 1

      if evaluacion == num:
        print(counter, ' iteraciones para el numero', str(num)+'.', 'Formula encontrada ->', formula)
        return [formula, counter]

In [58]:
# Aplicamos nuestro algoritmo mejorado al mismo numero
resultado_mejorado = search_formula_rand(rand_number)

628  iteraciones para el numero 52. Formula encontrada -> 6*9+2/1-4


(*)Calcula la complejidad del algoritmo 

Respuesta

#### La complejidad del algoritmo mejorado sigue siendo la misma, ya que solo modificamos el orden de iteracion de los bucles. De `O(1)` a `O(n*m)`

Según el problema (y tenga sentido), diseña un juego de datos de entrada aleatorios

Respuesta

In [62]:
# Como el algoritmo se basa en la aleatoriedad, vamos a comprobar si realmente mejora las iteraciones con un conjunto de datos aleatorios

# Conjunto de datos aleatorios dentro de nuestro espacio de soluciones
rand_numbers = [random.randint(min_value,max_value) for i in range(5)]

Aplica el algoritmo al juego de datos generado

Respuesta

In [65]:
# 5 numero aleatorios con el algoritmo sin mejorar
for num in rand_numbers:
  search_formula(num)

40531 iteraciones para el numero -35. Formula encontrada -> 2/1+3-5*8
42235 iteraciones para el numero -56. Formula encontrada -> 2/1+5-7*9
42115 iteraciones para el numero -47. Formula encontrada -> 2/1+5-6*9
14067 iteraciones para el numero 31. Formula encontrada -> 1+4*8-6/3
25089 iteraciones para el numero -49. Formula encontrada -> 1-6*9+8/2


In [66]:
# 5 numero aleatorios con el algoritmo mejorado
for num in rand_numbers:
  search_formula_rand(num)

1325  iteraciones para el numero -35. Formula encontrada -> 2+8/1-5*9
16  iteraciones para el numero -56. Formula encontrada -> 5+2-9*7/1
4293  iteraciones para el numero -47. Formula encontrada -> 6-7/1*8+3
156  iteraciones para el numero 31. Formula encontrada -> 4*8-3/1+2
426  iteraciones para el numero -49. Formula encontrada -> 5/1+2-7*8


Enumera las referencias que has utilizado(si ha sido necesario) para llevar a cabo el trabajo

Respuesta

#### Producto cartesiano: https://en.wikipedia.org/wiki/Cartesian_product#:~:text=In%20mathematics%2C%20specifically%20set%20theory,and%20a%20set%20of%20columns. <br>
#### Producto cartesiano de la libreria itertools: https://docs.python.org/3/library/itertools.html#itertools.product <br>
#### Permutaciones de la libreria itertools: https://docs.python.org/3/library/itertools.html#itertools.permutations <br>

Describe brevemente las lineas de como crees que es posible avanzar en el estudio del problema. Ten en cuenta incluso posibles variaciones del problema y/o variaciones al alza del tamaño

Respuesta

#### El estudio del problema se podría seguir estudiando y mejorando, ya que, aunque hemos mejorado bastante el número de iteraciones para encontrar una formula para un número dado, lo hemos resuelto con una tecnica metaheurísticas basada en la busqueda aleatoria. Y como toda técnica basada en la aleatoriedad, siempre hay un porcentaje de posibilidad de que el algoritmo no mejore las iteraciones. 
#### Se podría estudiar el uso de otros métodos metaheurísticos como la búsqueda tabu, que se basa en una técnica que intenta escapar de mínimos locales. 